aplib.net-demo/Assets/Scripts/Player/Movement.cs

```diff
@@ -0,0 +1,188 @@
+        _rigidbody.useGravity = !isOnSlope;
+
+        if (_isGrounded) {
+            _rigidbody.drag = _groundDrag; 
+            
+            if (isOnSlope) {
+                _rigidbody.AddForce(_maxSpeed * _acceleration * Time.fixedDeltaTime * 1.5f * directionOnSlope);
+                
+                // Apply gravity but perpendicular to the slope, to prevent sliding
+                _rigidbody.velocity += (_actualGravityScale - 1) * _gravity * Time.fixedDeltaTime * downHit.normal;
+            }
+            else {
+                _rigidbody.AddForce(_maxSpeed * _acceleration * Time.fixedDeltaTime * _horizontalVelocity.normalized);
+            }   
+        }
+        else {
+            _rigidbody.drag = 0;  
+            _rigidbody.AddForce(_maxSpeed * _acceleration * Time.fixedDeltaTime * _airMovementScale * _horizontalVelocity.normalized);
+
+            // Custom gravity for player while falling
+            _actualGravityScale = _rigidbody.velocity.y < 0 ? _fallingGravityScale : _normalGravityScale;
+            _rigidbody.velocity += (_actualGravityScale - 1) * _gravity * Time.fixedDeltaTime * Vector3.up;
+        }
+
+        if (isWallColliding) HandleWallCollision(wallNormal);
+        LimitVelocity(isOnSlope);
+    }
+
+    /// <summary>
+    /// Add force along the wall, to prevent sticking.
+    /// </summary>
+    /// <param name="wallNormal">The normal of the wall the player is walking against</param>
+    private void HandleWallCollision(Vector3 wallNormal) {
+        float angleMultiplier = Mathf.Clamp01(1.2f - Vector3.Angle(_horizontalVelocity, -wallNormal) / 90);
+        Vector3 directionOnWall = Vector3.ProjectOnPlane(_horizontalVelocity, wallNormal).normalized;
+        _rigidbody.AddForce(_maxSpeed * _acceleration * Time.fixedDeltaTime * angleMultiplier * directionOnWall);
+        Debug.DrawRay(transform.position, directionOnWall * 5, Color.blue);
+    }
+
+    /// <summary>
+    ///  Limit the player's velocity to the max speed.
+    ///  Also accounts for vertical velocity when on a slope.
+    /// </summary>
+    /// <param name="isOnWalkableSlope">Is the player on a walkable slope</param>
+    private void LimitVelocity(bool isOnWalkableSlope){
+        if (isOnWalkableSlope) {
+            if (_rigidbody.velocity.magnitude > _maxSpeed) _rigidbody.velocity = _rigidbody.velocity.normalized * _maxSpeed;   
+        } else {
+            Vector3 rigidbodyHorizontalVelocity = new(_rigidbody.velocity.x, 0, _rigidbody.velocity.z);
+            if (rigidbodyHorizontalVelocity.magnitude > _maxSpeed) {
+                Vector3 limitedVelocity = rigidbodyHorizontalVelocity.normalized * _maxSpeed;
+                _rigidbody.velocity = new Vector3(limitedVelocity.x, _rigidbody.velocity.y, limitedVelocity.z);
+            }
+        }
+        
+        // Draw a ray to visualize the player's velocity and direction
+        Debug.DrawRay(transform.position, _rigidbody.velocity, Color.red);
```

Comment:
Should this colour be defined somewhere so we have a uniform set of debug colour types?

(view rays are red, physics rays are blue etc etc that is an example but could be defined somewhere)